frontend/src/pages/register/form.tsx

```diff
@@ -1,76 +1,130 @@
-            handleInputChange={emailChangeHandler}
-          />
-          <Form.InputPassword
-            placeholder="Password"
-            value={password}
-            setPassword={passwordChangeHandler}
-          />
-          <Form.InputPassword
-            placeholder="Confirm password"
-            value={confirmPassword}
-            setPassword={confirmPasswordChangeHandler}
           />
-          <Form.Input
-            type="email"
+          {errors.email ? (
+            <Error>{errors.email.message}</Error>
+          ) : (
+            <Error>*</Error>
+          )}
+          <PasswordContainer>
+            <Input
+              {...register('password', {
+                required: 'This field is required',
+                minLength: {
+                  value: 8,
+                  message: 'Password must have at least 8 characters',
+                },
+                pattern: {
+                  value:
+                    /^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#&.()–[{}\]:;',?/*~$^+=<>])([^\s]){8,}$/i,
+                  message: 'Invalid password format',
+                },
+              })}
+              type={passwordVisible ? 'text' : 'password'}
+              placeholder="Password"
+            />
+            <SeenIcon
+              src={
+                passwordVisible
+                  ? './assets/icons/eyeclosed.png'
+                  : './assets/icons/eyeopen.png'
+              }
+              onClick={() => setPasswordVisible(!passwordVisible)}
+            />
+          </PasswordContainer>
+          {errors.password && (
+            <Form.Error>{errors.password.message}</Form.Error>
+          )}
+          <PasswordContainer>
+            <Input
+              {...register('confirmPassword', {
+                required: 'This field is required',
+                validate: (value) =>
+                  value === watch('password') || 'Passwords must match',
+              })}
+              type={passwordVisible ? 'text' : 'password'}
+              placeholder="Confirm password"
+            />
+            <SeenIcon
+              src={
+                passwordVisible
+                  ? './assets/icons/eyeclosed.png'
+                  : './assets/icons/eyeopen.png'
+              }
+              onClick={() => setPasswordVisible(!confirmPasswordVisible)}
+            />
+          </PasswordContainer>
+          {errors.confirmPassword && (
+            <Form.Error>{errors.confirmPassword.message}</Form.Error>
+          )}
+          <Input
+            {...register('firstName', { required: 'This field is required' })}
+            type="text"
             placeholder="First name"
-            value={firstName}
-            handleInputChange={firstNameChangeHandler}
           />
-          <Form.Input
-            type="email"
+          {errors.firstName && (
+            <Form.Error>{errors.firstName.message}</Form.Error>
+          )}
+          <Input
```

Comment:
Seems like very input should also have a way to display an error, so I think you can move error display to be part of the input component